README.md
# flexible_calendar

Flexible Calendar widget that makes custom a Calendar much easier! Based on the StatefulWidget, this widget works on both Android & iOS.

https://pub.dev/packages/flexible_calendar

# Installing
Add the following to your `pubspec.yaml` file:

``` diff
dependencies: 
  flexible_calendar: 1.0.0
  ```
  
# Simple Usage

After implement, you just use it like a normal child widget.
Like this

![alt](https://user-images.githubusercontent.com/78459285/126870689-21bb11d2-b1d8-4106-8059-261a709641f2.png)

## Screenshots

Choose a date

<img width="300" alt="Screen Shot 2021-07-24 at 20 56 31" src="https://user-images.githubusercontent.com/78459285/126870806-459e7b43-0e36-4ce3-8b02-7a4a2157b7f6.png">

Select date range

<img width="300" alt="Screen Shot 2021-07-24 at 20 56 31" src="https://user-images.githubusercontent.com/78459285/126870812-d16239aa-32a7-49c9-8df5-cd62de0f69eb.png">

# Custom Usage

There are several options that allow for more control:

Properties | Description
------------ | -------------
```initialStartDate``` | Initialize the start date of the calendar, you need to update it every time you select a date in the calendar.
```initialEndDate``` | Initialize the start date of the calendar. If use selection mode to select only one date, you can set ```initialEndDate = null```.
```startEndDateChange``` | Every time you click on the date, it will jump into this function for you to update the variable.
```minimumDate``` | Limit minimum date can choose.
```maximumDate``` | Limit maximum date can choose.
```isConvertDateOfWeekToVietnamese``` | Default date of week is "Mon, Tue, ...", please set **true** if you like it show "T2, T3, T4, ..."
```formatMonthTitle``` | Format title show top of calendar. Example: DateFormat("MM, yyyy") will show ```07, 2021```
```textPreMonthTitle``` | Set text pre title. Example: ```                      textPreMonthTitle: "Month", ``` will show ```Month 07, 2021```
```decorationOfSelectedDay``` | You can custom decoration of selected widget to suil design.
```colorOfDayInMonth``` | Set color of date in current month.
```colorDayInAnotherMonth``` | Set color of date in anothe month.
```colorOfSelectedDay``` | Set color of selected date.
```styleOfDay``` | Common style of all dates in the calendar.
```isSelectedOnlyOneDay``` | Enable mode select a date, default is select range.
```colorInRange``` | Set backgroundColor of range date, not contains StartDate and EndDate.









CHANGELOG.md
## [1.0.0] - 23/06/2021

* Fully working FlexibleCalendar; example included
## [1.0.1] - 24/06/2021

* Update Readme

test/flexible_calendar_test.dart
import 'package:flutter_test/flutter_test.dart';

import 'package:flexible_calendar/flexible_calendar.dart';

void main() {
  test('adds one to input values', () {
    final calculator = Calculator();
    expect(calculator.addOne(2), 3);
    expect(calculator.addOne(-7), -6);
    expect(calculator.addOne(0), 1);
  });
}


lib/flexible_calendar.dart
library flexible_calendar;

import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

class FlexibleCalendarView extends StatefulWidget {
  const FlexibleCalendarView(
      {Key? key,
      this.initialStartDate,
      this.initialEndDate,
      this.startEndDateChange,
      this.maximumDate,
      this.minimumDate,
      this.formatMonthTitle,
      this.textPreMonthTitle,
      this.isConvertDateOfWeekToVietnamese,
      this.formatTitleDateOfWeek,
      this.colorDayInAnotherMonth,
      this.colorOfSelectedDay,
      this.colorOfDayInMonth,
      this.styleOfDay,
      this.decorationOfSelectedDay,
      this.isSelectedOnlyOneDay,
      this.colorInRange,
      this.borderBackgroundStartDay,
      this.borderBackgroundEndDay})
      : super(key: key);

  final DateTime? minimumDate;
  final DateTime? maximumDate;
  final DateTime? initialStartDate;
  final DateTime? initialEndDate;
  final DateFormat? formatMonthTitle;
  final String? textPreMonthTitle;
  final bool? isConvertDateOfWeekToVietnamese;
  final String? formatTitleDateOfWeek;
  final Color? colorDayInAnotherMonth;
  final Color? colorOfSelectedDay;
  final Color? colorOfDayInMonth;
  final TextStyle? styleOfDay;
  final BoxDecoration? decorationOfSelectedDay;
  final bool? isSelectedOnlyOneDay;
  final Color? colorInRange;
  final BorderRadius? borderBackgroundStartDay;
  final BorderRadius? borderBackgroundEndDay;

  final Function(DateTime, DateTime?)? startEndDateChange;

  @override
  _FlexibleCalendarViewState createState() => _FlexibleCalendarViewState();
}

class _FlexibleCalendarViewState extends State<FlexibleCalendarView> {
  List<DateTime> dateList = <DateTime>[];
  DateTime? currentMonthDate = DateTime.now();
  DateTime? startDate;
  DateTime? endDate;

  @override
  void initState() {
    if (widget.initialStartDate != null) {
      currentMonthDate = widget.initialStartDate;
    }
    setListOfDate(currentMonthDate);
    if (widget.initialStartDate != null) {
      startDate = widget.initialStartDate;
    }
    if (widget.initialEndDate != null) {
      endDate = widget.initialEndDate;
    }
    super.initState();
  }

  @override
  void dispose() {
    super.dispose();
  }

  void setListOfDate(DateTime? monthDate) {
    dateList.clear();
    final DateTime newDate = DateTime(monthDate!.year, monthDate.month, 0);
    int previousMothDay = 0;
    if (newDate.weekday < 7) {
      previousMothDay = newDate.weekday;
      for (int i = 1; i <= previousMothDay; i++) {
        dateList.add(newDate.subtract(Duration(days: previousMothDay - i)));
      }
    }
    for (int i = 0; i < (42 - previousMothDay); i++) {
      dateList.add(newDate.add(Duration(days: i + 1)));
    }
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: Column(
        children: <Widget>[
          Padding(
            padding: const EdgeInsets.only(left: 8.0, right: 8.0, top: 4, bottom: 4),
            child: Row(
              mainAxisAlignment: MainAxisAlignment.center,
              children: <Widget>[
                Padding(
                  padding: EdgeInsets.all(8.0),
                  child: Container(
                    height: 38,
                    width: 38,
                    child: Material(
                      color: Colors.transparent,
                      child: InkWell(
                        borderRadius: const BorderRadius.all(Radius.circular(24.0)),
                        onTap: () {
                          setState(() {
                            currentMonthDate = DateTime(currentMonthDate!.year, currentMonthDate!.month, 0);
                            setListOfDate(currentMonthDate);
                          });
                        },
                        child: Icon(
                          Icons.keyboard_arrow_left,
                          color: Colors.grey,
                          size: 20,
                        ),
                      ),
                    ),
                  ),
                ),
                Center(
                  child: Text(
                    "${widget.textPreMonthTitle != null ? widget.textPreMonthTitle : "Tháng"} ${(widget.formatMonthTitle != null ? widget.formatMonthTitle : DateFormat('M, yyyy'))!.format(currentMonthDate!)}",
                    style: TextStyle(fontWeight: FontWeight.w500, fontSize: 17, color: Colors.black),
                  ),
                ),
                Padding(
                  padding: EdgeInsets.all(8.0),
                  child: Container(
                    height: 38,
                    width: 38,
                    child: Material(
                      color: Colors.transparent,
                      child: InkWell(
                        borderRadius: const BorderRadius.all(Radius.circular(24.0)),
                        onTap: () {
                          setState(() {
                            currentMonthDate = DateTime(currentMonthDate!.year, currentMonthDate!.month + 2, 0);
                            setListOfDate(currentMonthDate);
                          });
                        },
                        child: Icon(
                          Icons.keyboard_arrow_right,
                          color: Colors.grey,
                          size: 20,
                        ),
                      ),
                    ),
                  ),
                ),
              ],
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(right: 8, left: 8, bottom: 8),
            child: Row(
              children: getDaysNameUI(),
            ),
          ),
          Padding(
            padding: const EdgeInsets.only(right: 8, left: 8),
            child: Column(
              children: getDaysNoUI(),
            ),
          ),
        ],
      ),
    );
  }

  List<Widget> getDaysNameUI() {
    final List<Widget> listUI = <Widget>[];
    for (int i = 0; i < 7; i++) {
      listUI.add(
        Expanded(
          child: Center(
            child: Text(
              widget.isConvertDateOfWeekToVietnamese == true
                  ? convertDateOfWeekToVietnamese(
                      DateFormat(widget.formatTitleDateOfWeek != null ? widget.formatTitleDateOfWeek : 'EEE').format(dateList[i].toLocal()))
                  : DateFormat(widget.formatTitleDateOfWeek != null ? widget.formatTitleDateOfWeek : 'EEE').format(dateList[i].toLocal()),
              style: TextStyle(
                fontSize: 13,
                fontWeight: FontWeight.w500,
                color: Color(0xFF797E88),
              ),
            ),
          ),
        ),
      );
    }
    return listUI;
  }

  String convertDateOfWeekToVietnamese(String dayOfWeek) {
    switch (dayOfWeek.toUpperCase()) {
      case "MON":
        return "T2";
      case "TUE":
        return "T3";
      case "WED":
        return "T4";
      case "THU":
        return "T5";
      case "FRI":
        return "T6";
      case "SAT":
        return "T7";
      default:
        return "CN";
    }
  }

  List<Widget> getDaysNoUI() {
    final List<Widget> noList = <Widget>[];
    int count = 0;
    for (int i = 0; i < dateList.length / 7; i++) {
      final List<Widget> listUI = <Widget>[];
      for (int i = 0; i < 7; i++) {
        final DateTime date = dateList[count];
        listUI.add(
          Expanded(
            child: AspectRatio(
              aspectRatio: 1.0,
              child: Container(
                child: Stack(
                  children: <Widget>[
                    Padding(
                      padding: const EdgeInsets.only(top: 4, bottom: 4),
                      child: Material(
                        color: Colors.transparent,
                        child: Padding(
                          padding: EdgeInsets.only(top: 2, bottom: 2, left: isStartDateRadius(date) ? 4 : 0, right: isEndDateRadius(date) ? 4 : 0),
                          child: Container(
                            decoration: BoxDecoration(
                                color: startDate != null && endDate != null
                                    ? getIsItStartAndEndDate(date) || getIsInRange(date)
                                        ? widget.colorInRange ?? Color(0xFFE3ECFF)
                                        : Colors.transparent
                                    : Colors.transparent,
                                borderRadius: isStartDateRadius(date)
                                    ? widget.borderBackgroundStartDay ??
                                        BorderRadius.only(topLeft: Radius.circular(25), bottomLeft: Radius.circular(25))
                                    : isEndDateRadius(date)
                                        ? widget.borderBackgroundEndDay ??
                                            BorderRadius.only(topRight: Radius.circular(25), bottomRight: Radius.circular(25))
                                        : BorderRadius.circular(0)),
                          ),
                        ),
                      ),
                    ),
                    Material(
                      color: Colors.transparent,
                      child: InkWell(
                        borderRadius: const BorderRadius.all(Radius.circular(32.0)),
                        onTap: () {
                          if (widget.minimumDate != null) {
                            final DateTime newminimumDate =
                                DateTime(widget.minimumDate!.year, widget.minimumDate!.month, widget.minimumDate!.day - 1);
                            if (date.isAfter(newminimumDate)) {
                              onDateClick(date);
                            }
                          } else if (widget.maximumDate != null) {
                            final DateTime newmaximumDate = DateTime(widget.maximumDate?.year ?? DateTime.now().year,
                                widget.maximumDate?.month ?? DateTime.now().month, (widget.maximumDate?.day ?? DateTime.now().day) + 1);
                            if (date.isBefore(newmaximumDate)) {
                              onDateClick(date);
                            }
                          } else {
                            onDateClick(date);
                          }
                        },
                        child: Padding(
                          padding: const EdgeInsets.all(5),
                          child: Container(
                            decoration: getIsItStartAndEndDate(date)
                                ? widget.decorationOfSelectedDay ??
                                    BoxDecoration(
                                        color: Color(0xFF3874EF),
                                        borderRadius: const BorderRadius.all(Radius.circular(6)),
                                        boxShadow: <BoxShadow>[
                                          BoxShadow(color: Colors.grey.withOpacity(0.6), blurRadius: 4, offset: const Offset(0, 0)),
                                        ])
                                : BoxDecoration(
                                    color: Colors.transparent,
                                  ),
                            child: Center(
                              child: Text(
                                '${date.day}',
                                style:
                                    (widget.styleOfDay != null ? widget.styleOfDay : TextStyle(fontSize: 15, fontWeight: FontWeight.w500))!.copyWith(
                                  color: getIsItStartAndEndDate(date)
                                      ? widget.colorOfSelectedDay ?? Colors.white
                                      : currentMonthDate!.month == date.month
                                          ? widget.colorOfDayInMonth ?? Colors.black
                                          : widget.colorDayInAnotherMonth ?? Color(0xFF797E88),
                                ),
                              ),
                            ),
                          ),
                        ),
                      ),
                    ),
                    Positioned(
                      bottom: 9,
                      right: 0,
                      left: 0,
                      child: Container(
                        height: 6,
                        width: 6,
                        decoration: BoxDecoration(
                            color: DateTime.now().day == date.day && DateTime.now().month == date.month && DateTime.now().year == date.year
                                ? getIsInRange(date)
                                    ? Colors.white
                                    : Colors.blue
                                : Colors.transparent,
                            shape: BoxShape.circle),
                      ),
                    ),
                  ],
                ),
              ),
            ),
          ),
        );
        count += 1;
      }
      noList.add(Row(
        mainAxisAlignment: MainAxisAlignment.center,
        crossAxisAlignment: CrossAxisAlignment.center,
        mainAxisSize: MainAxisSize.min,
        children: listUI,
      ));
    }
    return noList;
  }

  bool getIsInRange(DateTime date) {
    if (startDate != null && endDate != null) {
      if (date.isAfter(startDate!) && date.isBefore(endDate!)) {
        return true;
      } else {
        return false;
      }
    } else {
      return false;
    }
  }

  bool getIsItStartAndEndDate(DateTime date) {
    if (startDate != null && startDate!.day == date.day && startDate!.month == date.month && startDate!.year == date.year) {
      return true;
    } else if (endDate != null && endDate!.day == date.day && endDate!.month == date.month && endDate!.year == date.year) {
      return true;
    } else {
      return false;
    }
  }

  bool isStartDateRadius(DateTime date) {
    if (startDate != null && startDate!.day == date.day && startDate!.month == date.month) {
      return true;
    } else if (date.weekday == 1) {
      return true;
    } else {
      return false;
    }
  }

  bool isEndDateRadius(DateTime date) {
    if (endDate != null && endDate!.day == date.day && endDate!.month == date.month) {
      return true;
    } else if (date.weekday == 7) {
      return true;
    } else {
      return false;
    }
  }

  void onDateClick(DateTime date) {
    if (widget.isSelectedOnlyOneDay == true) {
      startDate = date;
      endDate = null;
      setState(() {
        try {
          widget.startEndDateChange!(startDate!, endDate);
        } catch (_) {}
      });
    } else {
      if (startDate != null && endDate == null && date.isAfter(startDate!)) {
        endDate = date;
      } else {
        startDate = date;
        endDate = null;
      }
      setState(() {
        try {
          widget.startEndDateChange!(startDate!, endDate);
        } catch (_) {}
      });
    }
  }
}


example/lib/main.dart
import 'package:flexible_calendar/flexible_calendar.dart';
import 'package:flutter/material.dart';
import 'package:intl/intl.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);
  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  DateTime? timeChecking;
  DateTime? timeCheckout;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Container(),
      floatingActionButton: FloatingActionButton(
        onPressed: () {
          _showDialogCalendar(context);
        },
      ),
    );
  }

  _showDialogCalendar(BuildContext context) {
    showModalBottomSheet<void>(
        context: context,
        isScrollControlled: true,
        isDismissible: true,
        shape: RoundedRectangleBorder(
          borderRadius: BorderRadius.only(topRight: Radius.circular(15), topLeft: Radius.circular(15)),
        ),
        builder: (context) {
          return GestureDetector(
            onVerticalDragDown: (_) {},
            child: SingleChildScrollView(
              child: Container(
                height: 500,
                child: FlexibleCalendarView(
                  //limit start date
                  // minimumDate: isChecking || timeChecking == null ? DateTime.now() : timeChecking?.add(Duration(days: 1)),
                  //limit end date
                  // maximumDate: isChecking || timeChecking == null ? DateTime.now() : timeChecking?.add(Duration(days: 1)),
                  initialStartDate: timeChecking,
                  initialEndDate: timeCheckout,
                  startEndDateChange: (DateTime startDateData, DateTime? endDateData) {
                    setState(() {
                      timeChecking = startDateData;
                      timeCheckout = endDateData;

                    });
                  },
                  isConvertDateOfWeekToVietnamese: true,
                  formatMonthTitle: DateFormat("MM, yyyy"),
                  textPreMonthTitle: "Tháng",
                  decorationOfSelectedDay: BoxDecoration(color: Colors.blue, borderRadius: BorderRadius.circular(25)),
                  colorDayInAnotherMonth: Colors.grey,
                  colorInRange: Colors.blue.withOpacity(0.1),
                  borderBackgroundStartDay: BorderRadius.only(topLeft: Radius.circular(25), bottomLeft: Radius.circular(25)),
                  borderBackgroundEndDay: BorderRadius.only(topRight: Radius.circular(25), bottomRight: Radius.circular(25)),
                  formatTitleDateOfWeek: "EEE",
                  isSelectedOnlyOneDay: false,
                  colorOfDayInMonth: Colors.black,
                  colorOfSelectedDay: Colors.white,
                  styleOfDay: TextStyle(fontSize: 17, fontWeight: FontWeight.w500),
                ),
              ),
            ),
          );
        });
  }
}
